model/src/order.rs

```diff
@@ -258,21 +258,70 @@ impl OrderCreation {
-// pub struct OrderCancellation {
-//     pub uid: OrderUid,
-//     pub signature: Signature,
-// }
+/// An order cancellation as provided to the orderbook by the frontend.
+#[serde_as]
+#[derive(Eq, PartialEq, Clone, Copy, Debug, Deserialize, Serialize, Hash)]
+pub struct OrderCancellation {
+    // TODO - does it make sense to include the orderUid here?
+    pub order_uid: OrderUid,
+    pub signature: Signature,
+}
+
+impl OrderCancellation {
+    // keccak256("OrderCancellation(bytes orderUid)")
+    pub const ORDER_CANCELLATION_TYPE_HASH: [u8; 32] =
+        hex!("7b41b3a6e2b3cae020a3b2f9cdc997e0d420643957e7fea81747e984e47c88ec");
+
+    fn order_cancellation_digest(&self) -> [u8; 32] {
+        let mut hash_data = [0u8; 64];
+        hash_data[0..32].copy_from_slice(&Self::ORDER_CANCELLATION_TYPE_HASH);
+        hash_data[44..64].copy_from_slice(&signing::keccak256(&self.order_uid.0));
+        signing::keccak256(&hash_data)
+    }
+    // TODO - Many duplicated methods from OrderCreation, need EIP712 Signing Library
+    fn signing_digest_typed_data(&self, domain_separator: &DomainSeparator) -> [u8; 32] {
+        let mut hash_data = [0u8; 66];
+        hash_data[0..2].copy_from_slice(&[0x19, 0x01]);
+        hash_data[2..34].copy_from_slice(&domain_separator.0);
+        hash_data[34..66].copy_from_slice(&self.order_cancellation_digest());
+        signing::keccak256(&hash_data)
+    }
+
+    fn signing_digest_message(&self, domain_separator: &DomainSeparator) -> [u8; 32] {
+        let mut hash_data = [0u8; 92];
+        hash_data[0..28].copy_from_slice(b"\x19Ethereum Signed Message:\n64");
+        hash_data[28..60].copy_from_slice(&domain_separator.0);
+        hash_data[60..92].copy_from_slice(&self.order_cancellation_digest());
+        signing::keccak256(&hash_data)
+    }
+
+    fn signing_digest(&self, domain_separator: &DomainSeparator) -> [u8; 32] {
+        // This is fallback for wallets that don't support EIP712
+        if self.signature.v & 0x80 == 0 {
+            self.signing_digest_typed_data(domain_separator)
+        } else {
+            self.signing_digest_message(domain_separator)
+        }
+    }
+
+    // returns owner if valid.
+    pub fn validate_signature(&self, domain_separator: &DomainSeparator) -> Option<H160> {
+        let v = self.signature.v & 0x1f;
+        // Does it make sense to include just order_uid here?
+        let message = self.signing_digest(domain_separator);
+        let recovery = Recovery::new(message, v as u64, self.signature.r, self.signature.s);
+        let (signature, recovery_id) = recovery.as_signature()?;
+        signing::recover(&message, &signature, recovery_id).ok()
+    }
+}
 
 /// An order as provided to the orderbook by the frontend.
 #[serde_as]
 #[derive(Eq, PartialEq, Clone, Debug, Deserialize, Serialize, Hash)]
 #[serde(rename_all = "camelCase")]
 pub struct OrderMetaData {
     pub creation_date: DateTime<Utc>,
+    pub cancellation_date: Option<DateTime<Utc>>,
```

Comment:
Does this need to become part of the OpenApi? I'm also wondering if we even need to differentiate between "on-chain" invalidated and "soft" cancelled field (we could reuse `invalidated` on this struct). If we do introduce a new field we should probably make sure the naming reflects these two differences in a more obvious way and the data returned is roughly in the same format (e.g. return a timestamp for both).

Reply:
I think we can simply go with making invalidated an optional date time and use the same field for both.

Reply:
Right now we don't store the timestamp of the block that carried the invalidated event, but it's something we can add.

Reply:
From the discussion above, I thought we agreed to reuse the existing `invalidated` field instead of creating a new field on the OrderMetaData struct.

Reply:
Right. I only included this because I thought it might be useful to pass whatever metadata we have on the order when queried. Of course we can still use invalidated, but there is a subtle loss of information (that being the actual time when it was invalidated). However, I agree, its not all that necessary and the boolean should suffice.